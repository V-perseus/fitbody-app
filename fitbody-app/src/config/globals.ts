import { Dimensions, TextProps } from 'react-native'
import styles from './styles'
import string from './strings'
import { FB_CURRENT_ENV, FB_ENVIRONMENTS } from '../../env'

export const globals = {
  // Styles & Colors
  styles: {
    ...styles,
  },
  // Fonts
  fonts: {
    ...styles.fonts,
  },
  // Header Styles
  header: {
    icons: {
      marginRight: 15,
      marginLeft: 15,
    },
    headerTitleStyle: <TextProps>{
      fontSize: 16,
      fontFamily: styles.fonts.primary.semiboldStyle.fontFamily,
      textAlign: 'center',
    },
    headerTruncatedBackTitle: true,
  },
  // Standard String
  strings: {
    ...string,
  },
  // Background Images
  bgImage: {
    imageStyle: { resizeMode: 'cover' },
    style: {
      width: Dimensions.get('window').width,
      height: Dimensions.get('window').height,
    },
  },
  // Window dimensions
  window: {
    width: Dimensions.get('window').width,
    height: Dimensions.get('window').height,
  },
  // Button click targeting
  touchableOpacity: {
    activeOpacity: 0.8,
    hitSlop: { top: 5, right: 5, left: 5, bottom: 5 },
  },
  fbEnvironments: {
    ...FB_ENVIRONMENTS,
  },
  currentEnv: FB_CURRENT_ENV,
  get apiBase(): string {
    return this.fbEnvironments[this.currentEnv]
  },
}

export default globals
