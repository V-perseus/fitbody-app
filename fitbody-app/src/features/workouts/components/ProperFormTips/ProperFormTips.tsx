import React, { useEffect, useState, memo, useCallback, useRef } from 'react'
import { View, Text, StyleSheet } from 'react-native'
import { useFocusEffect } from '@react-navigation/native'
import { ms, vs } from 'react-native-size-matters/extend'
import { Audio, AVPlaybackStatus } from 'expo-av'

import BottomUpPanel from '../../../../components/BottomUpPanel'
import { ButtonOpacity } from '../../../../components/Buttons/ButtonOpacity'

import SoundOnIcon from '../../../../../assets/images/svg/icon/40px/audio-on.svg'
import SoundOffIcon from '../../../../../assets/images/svg/icon/40px/audio-off.svg'
import CloseIcon from '../../../../../assets/images/svg/icon/24px/close.svg'
import globals from '../../../../config/globals'

interface IProperFormTipsProps {
  content: string
  audioUrl: string
  autoplayEnabled: boolean
  onOpenStateChange?: (open: boolean) => void
  open?: boolean
  disabled: boolean
}
function ProperFormTips({ content, audioUrl, autoplayEnabled, onOpenStateChange, open = false, disabled = false }: IProperFormTipsProps) {
  const [isOpen, setIsOpen] = useState<boolean>(open)
  const [isPlaying, setIsPlaying] = useState<boolean>(false)
  const [sound, setSound] = useState<Audio.Sound | null>(null)

  const handleStatusChange = (status: AVPlaybackStatus) => {
    if (status.isLoaded && status.didJustFinish) {
      setIsPlaying(false)
      sound?.unloadAsync()
    }
  }

  const play = async () => {
    if (isPlaying) {
      await stopAndUnload()
    }
    if (audioUrl) {
      const { sound: s } = await Audio.Sound.createAsync({ uri: audioUrl }, {}, handleStatusChange)
      setSound(s)
      setIsPlaying(true)
      await s.playAsync()
    }
  }

  const stop = async () => {
    try {
      setIsPlaying(false)
      await sound?.stopAsync()
    } catch (error) {}
  }

  const init = async () => {
    if (autoplayEnabled && audioUrl) {
      play()
    } else if (isPlaying) {
      stopAndUnload()
    }
  }

  const stopAndUnload = async () => {
    try {
      if (isPlaying) {
        await stop()
      }
      await sound?.unloadAsync()
    } catch (error) {}
  }

  useFocusEffect(
    useCallback(() => {
      // init()
    }, [audioUrl, autoplayEnabled]),
  )

  useFocusEffect(
    useCallback(() => {
      return sound
        ? () => {
            sound.unloadAsync()
          }
        : undefined
    }, [sound, audioUrl]),
  )

  useEffect(() => {
    setIsOpen(open)
  }, [open])

  function handleAnimStart(val: boolean) {
    disabled ? null : onOpenStateChange?.(val)
  }

  function renderHeader() {
    return (
      <View style={styles.headerContainer}>
        {isOpen && (
          <View style={styles.headerButtonContainer}>
            {/* {isPlaying ? (
              <ButtonOpacity onPress={stop} testID="stop">
                <SoundOffIcon color={globals.styles.colors.colorBlack} />
              </ButtonOpacity>
            ) : (
              <ButtonOpacity onPress={play} testID="play">
                <SoundOnIcon color={globals.styles.colors.colorPink} />
              </ButtonOpacity>
            )} */}
          </View>
        )}
        <Text
          style={{
            fontFamily: globals.fonts.primary.boldStyle.fontFamily,
            fontSize: 18,
            flexGrow: 1,
            textAlign: 'center',
            marginTop: -5,
            color: disabled ? globals.styles.colors.colorGrayDark : globals.styles.colors.colorBlack,
          }}>
          Proper Form Tips
        </Text>
        {isOpen && (
          <View style={styles.headerButtonContainer}>
            <CloseIcon style={{ marginLeft: 'auto' }} color={globals.styles.colors.colorBlack} />
          </View>
        )}
      </View>
    )
  }

  function renderContent() {
    return (
      <View style={styles.contentContainer}>
        <View style={{ paddingHorizontal: 48 }}>
          <Text style={styles.tipText}>{content}</Text>
        </View>
      </View>
    )
  }

  return (
    <BottomUpPanel
      containerStyle={styles.panelContainer}
      isOpen={disabled ? false : isOpen}
      customHeader={renderHeader}
      content={renderContent}
      startHeight={vs(150)}
      topEnd={globals.window.height - ms(340)}
      onAnimationStart={handleAnimStart}
      contentContainerStyle={{ height: 156, borderWidth: 0 }}
      disabled={disabled}
      fadeScrollContent={true}
      showVerticalScrollIndicator={true}
      bounces={true}
    />
  )
}

export default memo(ProperFormTips)

const styles = StyleSheet.create({
  headerContainer: { flexDirection: 'row', paddingHorizontal: 16, height: 49, backgroundColor: globals.styles.colors.colorWhite },
  headerButtonContainer: { height: 48, width: 48 },
  panelContainer: {
    bottom: 20,
    zIndex: 0,
    height: globals.window.height - ms(440),
  },
  contentContainer: { marginTop: 5, marginBottom: 40 },
  tipText: { fontSize: 14, fontFamily: globals.fonts.primary.style.fontFamily, color: globals.styles.colors.colorBlack },
})
