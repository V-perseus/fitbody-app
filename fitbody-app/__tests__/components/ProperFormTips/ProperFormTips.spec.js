import React from 'react'
import { act, cleanup, fireEvent, render, waitFor } from '@testing-library/react-native'
import { NavigationContext } from '@react-navigation/native'

import ProperFormTips from '../../../src/features/workouts/components/ProperFormTips/ProperFormTips'
import globals from '../../../src/config/globals'

jest.mock('expo-av', () => {
  const { Audio } = jest.requireActual('expo-av')
  Audio.Sound.createAsync = jest.fn(() =>
    Promise.resolve({ sound: { playAsync: jest.fn(), unloadAsync: jest.fn(), stopAsync: jest.fn() } }),
  )
  return { Audio }
})

afterEach(cleanup)
// fake NavigationContext value data
const navContext = {
  isFocused: () => true,
  // addListener returns an unscubscribe function.
  addListener: jest.fn(() => jest.fn()),
}

const renderProps = (props = {}) => ({
  content: 'this is a tip',
  audioUrl: 'https://www.google.com',
  autoplayEnabled: false,
  onOpenStateChange: jest.fn(),
  open: false,
  disabled: false,
  ...props,
})

describe('ProperFormTips', () => {
  it('renders snapshot correctly', () => {
    const tree = render(
      <NavigationContext.Provider value={navContext}>
        <ProperFormTips {...renderProps()} />
      </NavigationContext.Provider>,
    ).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render header and content text', () => {
    const { queryByText } = render(
      <NavigationContext.Provider value={navContext}>
        <ProperFormTips {...renderProps()} />
      </NavigationContext.Provider>,
    )
    expect(queryByText(/Proper Form Tips/)).not.toBeNull()
    expect(queryByText(/this is a tip/)).not.toBeNull()
  })

  // audio toggle buttons have been temp disabled
  // it('should show play buttons when in opened state', async () => {
  //   const { queryByTestId } = render(
  //     <NavigationContext.Provider value={navContext}>
  //       <ProperFormTips {...renderProps({ open: true })} />
  //     </NavigationContext.Provider>,
  //   )
  //   expect(queryByTestId('play')).not.toBeNull()
  //   expect(queryByTestId('stop')).toBeNull()
  // })

  // it('should hide play button when menu is closed', async () => {
  //   const { queryByTestId } = render(
  //     <NavigationContext.Provider value={navContext}>
  //       <ProperFormTips {...renderProps({ open: false })} />
  //     </NavigationContext.Provider>,
  //   )
  //   expect(queryByTestId('play')).toBeNull()
  //   expect(queryByTestId('stop')).toBeNull()
  // })

  // audio toggle buttons have been temp disabled
  // it('should toggle play/stop when play is pressed', async () => {
  //   const { queryByTestId } = render(
  //     <NavigationContext.Provider value={navContext}>
  //       <ProperFormTips {...renderProps({ open: true, autoplayEnabled: true })} />
  //     </NavigationContext.Provider>,
  //   )
  //   const playBtn = queryByTestId('play')
  //   expect(playBtn).not.toBeNull()
  //   act(() => {
  //     fireEvent.press(playBtn)
  //   })
  //   await waitFor(() => queryByTestId('stop'))
  //   expect(queryByTestId('play')).toBeNull()
  //   expect(queryByTestId('stop')).not.toBeNull()
  // })

  it('should be properly disabled', async () => {
    const { queryByTestId, queryByText } = render(
      <NavigationContext.Provider value={navContext}>
        <ProperFormTips {...renderProps({ disabled: true })} />
      </NavigationContext.Provider>,
    )
    const playBtn = queryByTestId('play')
    expect(playBtn).toBeNull()
    expect(queryByText(/Proper Form Tips/)).toHaveStyle({ color: globals.styles.colors.colorGrayDark })
  })
})
