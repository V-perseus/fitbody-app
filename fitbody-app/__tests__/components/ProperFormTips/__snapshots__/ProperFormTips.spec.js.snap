// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProperFormTips renders snapshot correctly 1`] = `
<View
  pointerEvents="box-none"
  style={
    [
      {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "flex-end",
        "left": 0,
        "position": "absolute",
        "right": 0,
      },
      {
        "bottom": 20,
        "height": 715.4492753623188,
        "zIndex": 0,
      },
    ]
  }
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "transparent",
        "bottom": 0,
        "height": 1334,
        "left": 0,
        "paddingBottom": 856.0289855072464,
        "position": "absolute",
        "right": 0,
        "transform": [
          {
            "translateY": 1110.674107142857,
          },
          {
            "translateX": 0,
          },
        ],
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={18}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          {
            "alignItems": "center",
            "backgroundColor": "#ffffff",
            "borderTopLeftRadius": 54.34782608695652,
            "borderTopRightRadius": 54.34782608695652,
            "display": "flex",
            "elevation": 7,
            "justifyContent": "center",
            "shadowColor": "#000000",
            "shadowOffset": {
              "height": 0,
              "width": 0,
            },
            "shadowOpacity": 0.17,
            "shadowRadius": 10,
            "width": 750,
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#dedede",
              "borderRadius": 2.5,
              "height": 5,
              "justifyContent": "center",
              "marginVertical": 10,
              "width": 32,
            }
          }
        />
        <View
          style={
            {
              "backgroundColor": "#ffffff",
              "flexDirection": "row",
              "height": 49,
              "paddingHorizontal": 16,
            }
          }
        >
          <Text
            style={
              {
                "color": "#333333",
                "flexGrow": 1,
                "fontFamily": "OpenSans-Bold",
                "fontSize": 18,
                "marginTop": -5,
                "textAlign": "center",
              }
            }
          >
            Proper Form Tips
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        {
          "borderWidth": 0,
          "height": 156,
        }
      }
    >
      <RCTScrollView
        bounces={true}
        contentContainerStyle={
          {
            "backgroundColor": "#ffffff",
            "flexGrow": 1,
          }
        }
        indicatorStyle="black"
        scrollEnabled={false}
        scrollEventThrottle={16}
        showsHorizontalScrollIndicator={false}
        showsVerticalScrollIndicator={true}
      >
        <View>
          <View
            style={
              {
                "marginBottom": 40,
                "marginTop": 5,
              }
            }
          >
            <View
              style={
                {
                  "paddingHorizontal": 48,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#333333",
                    "fontFamily": "OpenSans",
                    "fontSize": 14,
                  }
                }
              >
                this is a tip
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
      <BVLinearGradient
        colors={
          [
            16777215,
            4294967295,
          ]
        }
        endPoint={
          {
            "x": 0.5,
            "y": 1,
          }
        }
        locations={null}
        pointerEvents="none"
        startPoint={
          {
            "x": 0.5,
            "y": 0,
          }
        }
        style={
          {
            "borderRadius": 10,
            "borderWidth": 0,
            "bottom": 0,
            "height": 10,
            "position": "absolute",
            "width": 750,
          }
        }
      />
      <BVLinearGradient
        colors={
          [
            4294967295,
            16777215,
          ]
        }
        endPoint={
          {
            "x": 0.5,
            "y": 1,
          }
        }
        locations={null}
        pointerEvents="none"
        startPoint={
          {
            "x": 0.5,
            "y": 0,
          }
        }
        style={
          {
            "borderRadius": 10,
            "borderWidth": 0,
            "height": 10,
            "position": "absolute",
            "top": 0,
            "width": 750,
          }
        }
      />
    </View>
  </View>
</View>
`;
